### DISCUSSION TITLE
Translation not working when deploying with Deployer

### DISCUSSION THREAD

**Discussion by muyncky** [PrestaShop/PrestaShop#33398] [Ideas]
When deploying a PrestaShop 1.7.6.2 application with Deployer, the translations do not work. It seems that referring to translation files does not work and/or cache is filled correctly.

## Deployer

When deploying with [Deployer](https://deployer.org/) ([deployphp/deployer](https://github.com/deployphp/deployer/)) the code base is separated into `shared` files and `release` files. The shared files are the assets that will be updated by using the application, the release files are basically the repository files. So this will follow, more or less, the .gitignore list. And example file structure of a production situation:

```

.
├── current >> releases/3 # symbolic link to the latest release
├── releases
│   ├── 1
│   ├── 2
│   └── 3
│       ├── admin625kvim5d
│       ├── app
│       ├── bin
│       ├── cache
│       ├── classes
│       ├── config
│       ├── controllers
│       ├── modules
│       ├── override
│       ├── recipe
│       ├── src
│       ├── themes
│       ├── vendor
│       └── webservice
└── shared
    ├── docs
    │   ├── csv_import
    │   ├── docker
    │   ├── licences
    │   └── server_config
    ├── download
    ├── img
    │   ├── a
    │   ├── admin
    │   ├── c
    │   ├── cms
    │   ├── co
    │   ├── footer
    │   ├── genders
    │   ├── jquery-ui
    │   ├── l
    │   ├── m
    │   ├── os
    │   ├── p
    │   ├── s
    │   ├── scenes
    │   ├── st
    │   ├── su
    │   ├── t
    │   └── tmp
    ├── js
    │   ├── admin
    │   ├── cropper
    │   ├── jquery
    │   ├── tiny_mce
    │   └── vendor
    ├── localization
    │   └── CLDR
    ├── mails
    │   ├── en
    │   ├── nl
    │   └── themes
    ├── pdf
    ├── tools
    │   ├── parser_sql
    │   └── profiling
    ├── translations
    │   ├── cldr
    │   └── export
    ├── upload
    │   ├── dtporder
    │   └── todos
    └── var
        ├── cache
        └── logs

```

### .htaccess

In the root, next to `current`, `releases` and `shared` is an .htaccess file pointing to the `current`, `current` in turn, points to `release/3`.

```
Options -Indexes

<IfModule mod_rewrite.c>
    RewriteEngine On 
    RewriteRule ^(.*)$ current/$1 [L]
</IfModule>

```

### Cause of the problem

The `var/cache` and `app/Resources/translations` folders are placed in shared, because a PrestaShop-user with a Translator profile might update translations from the back office, and this should not be overwritten by a new deployment (update of latest git version).

**What works (manually)**

When I manually move files from local dev environment `/var/cache/dev/translations/` to  `/shared/var/cache/dev/translations`, the translations are back -- for the most part. A module, for example, _ps_featuredproducts_ is not translated when moving these cache files.

The files I find locally are these:

```
.
├── catalogue.nl-NL.L8dqxxF.php
├── catalogue.nl-NL.L8dqxxF.php.meta
├── catalogue.nl-NL.NXhscRe.php
└── catalogue.nl-NL.NXhscRe.php.meta
```

The files in the deployed version:

```
.
├── catalogue.nl-NL.L8dqxxF.php
└── catalogue.nl-NL.L8dqxxF.php.meta
```

But when the cache is filled on the deployed version, the missing files with `NXhscRe` token are missing. So when moving them from local to deployed (with FTP), the translations are back.

Since 1.7.6.2, warming up cache is possible with the command `php bin/console cache:warmup`, but stil those files are not added (a lot of other languages are added). But that token is not matching. 

### Translation in the back office

Going to translation in the back office I get the message "_There are no translation file available._"

## Importance

I believe that it is important that PrestaShop can be deployed with Deployer. We can achief deployment with 0 downtime. Deployer is written in php so PrestaShop developers can extend Deployer recipes (set of instruction) for perfect deployment of PrestaShop. I think that PrestaShop 1.7 is well on it's way to be perfectly compatible with Deployer becuase of a new set of cli commands (`php bin/console`).

I have a PrestaShop 1.7 recipe ready here, I want to share it once we find a solution for the translations.

Thanks for the help!

**Additional information**
PrestaShop version: 1.7.6.2
PHP version: 7.x

**Deployer recipe**

```
<?php
namespace Deployer;

require 'recipe/common.php';

// Project name
set('application', 'shop.example.org');


// Project repository
set('repository', 'git@github.com:username/example.org.git');

// [Optional] Allocate tty for git clone. Default value is false.
set('git_tty', true); 

set('symfony_env', 'prod');

set('theme', 'child-theme');

// Shared files/dirs between deploys 
set('shared_files', [
    'app/config/parameters.php',
    '.htaccess',
    '.user.ini',
]);

set('shared_dirs', [
      'docs',
      'download',
      'img',
      'js',
      'localization',
      'mails/themes',
      'modules/ps_imageslider/images',
      'pdf',
      'tools',
      'translations',
      'upload',
      'var',
      'app/logs',
      'app/Resources/translations/default',
      'app/Resources/translations/nl-NL',
      'modules/blockreassurance/img',
      'modules/blockreassurance/views/img',
      'themes/classic/translations/nl-NL',
      'themes/child-theme/translations/nl-NL',
    ]
);

// Writable dirs by web server 
set('writable_dirs', [
      'app',
      'config',
      'config/xml',
      'download',
      'img',
      'mails/themes',
      'pdf',
      'translations',
      'upload',
      'var',
      'app/Resources/translations/default',
      'app/Resources/translations/nl-NL',
      'modules/blockreassurance/img',
      'modules/blockreassurance/views/img',
      'themes/classic/translations/nl-NL',
      'themes/child-theme/translations/nl-NL',
  ]
);

set('allow_anonymous_stats', false);

// Hosts
host('example')
    ->stage('dev')
    ->set('deploy_path', '/home/user/domains/shop.example.org/public_html')
    ->set('branch', 'dev');

task('prestashop:remove_cache', function(){
    run('cd {{deploy_path}}/shared/var/ && rm -rf *');
});

task('prestashop:cache_warmup', function(){
    // warmup current mode
    // writeln(run('cd {{release_path}}/ && php bin/console cache:warmup'));
    // warmup both
    writeln(run('cd {{release_path}}/ && php bin/console cache:warmup --env=dev'));
    writeln(run('cd {{release_path}}/ && php bin/console cache:warmup --env=prod'));
});

task('prestashop:enable_theme', function(){
    writeln(run('cd {{release_path}}/ && php bin/console prestashop:theme:enable {{theme}}'));
});

// Tasks
desc('Deploy your project');
task('deploy', [
    'deploy:info',
    'deploy:prepare',
    'deploy:lock',
    'deploy:release',
    'deploy:update_code',
    'deploy:shared',
    'deploy:writable',
    // 'deploy:vendors',
    'deploy:clear_paths',
    'deploy:symlink',
    'prestashop:enable_theme',
    'deploy:unlock',
    'prestashop:remove_cache',
    'prestashop:cache_warmup',
    'cleanup',
    'success'
]);

// [Optional] If deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');

```

**Comment by matks** [comment]
Sorry I could not reply sooner, we've been quite busy with https://build.prestashop.com/news/critical-security-vulnerability-in-prestashop-modules/ 😓

**Comment by muyncky** [comment]
@matks no worries, our sites are secured from that bug! Thanks for alarming me in every possible channel ;-)

Good luck!

**Comment by giacomo-secchi** [comment]
Some suggestion to solve the bug by myself?
I've developed a Deployer recipe and this seems to be the biggest problem… 

Please HELP

**Comment by PierreRambaud** [comment]
Did you try to remove cache after deploying?

**Comment by muyncky** [comment]
@PierreRambaud 

Yes.

`run('cd {{deploy_path}}/shared/var/ && rm -rf *');` is responsible for removing the /var folder. Emptying cache/warmin up with Symfony cli also doesn't work. Using `Tools`-class to empty various things doesn't either.

The thing is with language cache files is that emptying is not a problem, the [filling seems to fail](https://github.com/PrestaShop/PrestaShop/issues/17011#issue-544565240):

> When I manually move files from local dev environment /var/cache/dev/translations/ to /shared/var/cache/dev/translations, the translations are back -- for the most part. A module, for example, ps_featuredproducts is not translated when moving these cache files.

**Comment by PierreRambaud** [comment]
Are you sure you need to remove the `var` directory in the `shared` one?

**Comment by muyncky** [comment]
@PierreRambaud 

Emptying `var` should removable at any time. It contains cached translations and configs.

It is an environment specific folder, and therefor is not part of the versioned files. So, moving it to `writable_dirs` is not an option.

**Comment by PierreRambaud** [comment]
Yes I know about the `var` directory, I mean:
```
run('cd {{deploy_path}}/shared/var/ && rm -rf *');
// Could also be
run('rm -rf {{deploy_path}}/shared/var/*');
```

Are you sure `{{deploy_path}}/shared/var/` is the right path?

**Comment by muyncky** [comment]
That is a way too, I guess cleaner. I also tried this manually, it doesn't solve the translation issue. This is not the issue.

This part is the issue:
```
.
├── catalogue.nl-NL.L8dqxxF.php
└── catalogue.nl-NL.L8dqxxF.php.meta
```

After removing the var or cache is emptied, these files are put there, and they are not used/not correct.

**Comment by PierreRambaud** [comment]
@muyncky Could you provide me a full example of the recipe you use, the way you build, and how you deploy?
I need to know if the problem comes from the Deployer process or if it's a Core's problem :(

**Comment by muyncky** [comment]
**Recipe**

```
<?php
namespace Deployer;

require 'recipe/common.php';

// Project name
set('application', 'shop.example.org');


// Project repository
set('repository', 'git@github.com:username/example.org.git');

// [Optional] Allocate tty for git clone. Default value is false.
set('git_tty', true); 

set('symfony_env', 'prod');

set('theme', 'child-theme');

// Shared files/dirs between deploys 
set('shared_files', [
    'app/config/parameters.php',
    '.htaccess',
    '.user.ini',
]);

set('shared_dirs', [
      'docs',
      'download',
      'img',
      'js',
      'localization',
      'mails/themes',
      'modules/ps_imageslider/images',
      'pdf',
      'tools',
      'translations',
      'upload',
      'var',
      'app/logs',
      'app/Resources/translations/default',
      'app/Resources/translations/nl-NL',
      'modules/blockreassurance/img',
      'modules/blockreassurance/views/img',
      'themes/classic/translations/nl-NL',
      'themes/child-theme/translations/nl-NL',
    ]
);

// Writable dirs by web server 
set('writable_dirs', [
      'app',
      'config',
      'config/xml',
      'download',
      'img',
      'mails/themes',
      'pdf',
      'translations',
      'upload',
      'var',
      'app/Resources/translations/default',
      'app/Resources/translations/nl-NL',
      'modules/blockreassurance/img',
      'modules/blockreassurance/views/img',
      'themes/classic/translations/nl-NL',
      'themes/child-theme/translations/nl-NL',
  ]
);

set('allow_anonymous_stats', false);

// Hosts
host('example')
    ->stage('dev')
    ->set('deploy_path', '/home/user/domains/shop.example.org/public_html')
    ->set('branch', 'dev');

task('prestashop:remove_cache', function(){
    run('cd {{deploy_path}}/shared/var/ && rm -rf *');
});

task('prestashop:cache_warmup', function(){
    // warmup current mode
    // writeln(run('cd {{release_path}}/ && php bin/console cache:warmup'));
    // warmup both
    writeln(run('cd {{release_path}}/ && php bin/console cache:warmup --env=dev'));
    writeln(run('cd {{release_path}}/ && php bin/console cache:warmup --env=prod'));
});

task('prestashop:enable_theme', function(){
    writeln(run('cd {{release_path}}/ && php bin/console prestashop:theme:enable {{theme}}'));
});

// Tasks
desc('Deploy your project');
task('deploy', [
    'deploy:info',
    'deploy:prepare',
    'deploy:lock',
    'deploy:release',
    'deploy:update_code',
    'deploy:shared',
    'deploy:writable',
    // 'deploy:vendors',
    'deploy:clear_paths',
    'deploy:symlink',
    'prestashop:enable_theme',
    'deploy:unlock',
    'prestashop:remove_cache',
    'prestashop:cache_warmup',
    'cleanup',
    'success'
]);

// [Optional] If deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');
```

There is nothing to build in terms JavaScript libraries.

**Deployment**

I run `dep deploy {{env}}`, like `dep deploy dev` or `dep deploy prod`.

**Comment by sudarshann** [comment]
Any Solutions on this ?

**Comment by muyncky** [comment]
@sudarshann no.

I haven't tested with 1.7.7.0 though.

**Comment by bartoll** [comment]
@muyncky have you had a chance to test it? Does it work now?

**Comment by muyncky** [comment]
@bartoll no haven't had the time yet. Also, we only have one project in staging that is built with 1.7.7.x.

@PierreRambaud are all translations stored in the database since 1.7.7.0?

**Comment by PierreRambaud** [comment]
Same haven't had the time yet to test it :( 

@muyncky as far as I know, no, some are stores in files, other in the database, it depends :sweat_smile:

**Comment by louisbels** [comment]
Hello, 
I got the same issue with the 1.7.8.7 so I removed translations folders from the shared_dirs and put them in the copy_dirs with the copy_dirs recipe.
This may fail when we unlock a deployment but it seems to work.
```
require 'recipe/deploy/copy_dirs.php';
set('copy_dirs', [
    'app/Resources/translations/es-ES',
    'app/Resources/translations/fr-FR'
]);
add('shared_dirs', [
    //'app/Resources/translations/es-ES',
    //'app/Resources/translations/fr-FR',
    'var/logs',
    'var/cache'
]);
```
